ldap/basic_ldap.py

```diff
@@ -0,0 +1,120 @@
+import sys
+
+import ldap  # pylint: disable=import-error
+from flask import current_app, jsonify, request
+from flask_cors import cross_origin
+
+from alerta.auth.utils import create_token, get_customers
+from alerta.exceptions import ApiError
+from alerta.models.permission import Permission
+from alerta.models.user import User
+from alerta.utils.audit import auth_audit_trail
+
+from . import auth
+#custom
+
+@auth.route('/auth/login', methods=['OPTIONS', 'POST'])
+@cross_origin(supports_credentials=True)
+def login():
+    # Allow LDAP server to use a self signed certificate
+    if current_app.config['LDAP_ALLOW_SELF_SIGNED_CERT']:
+        ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_ALLOW)
+
+    # Retrieve required fields from client request
+    try:
+        login = request.json.get('username', None) or request.json['email']
+        password = request.json['password']
+    except KeyError:
+        raise ApiError("must supply 'username' and 'password'", 401)
+
+    try:
+        if '\\' in login:
+            domain, username = login.split('\\')
+            email = ''
+            email_verified = False
+        elif '@' in login:
+            username, domain = login.split('@')
+            email = login
+            email_verified = True
+        else:
+            try:
+              username, domain = login, current_app.config['LDAP_DEFAULT_DOMAIN']
+              email = ''
+              email_verified = False
+            except KeyError:
+              raise ApiError('LDAP_DEFAULT_DOMAIN has to be defined', 401)
+    except ValueError:
+        raise ApiError('expected username with domain', 401)
+
+    # Validate LDAP domain
+    if domain not in current_app.config['LDAP_DOMAINS']:
+        raise ApiError('unauthorized domain', 403)
+
+    userdn = current_app.config['LDAP_DOMAINS'][domain] % username
+
+    # Attempt LDAP AUTH
+    try:
+        trace_level = 2 if current_app.debug else 0
+        ldap_connection = ldap.initialize(current_app.config['LDAP_URL'], trace_level=trace_level)
+        ldap_connection.simple_bind_s(userdn, password)
+    except ldap.INVALID_CREDENTIALS:
+        raise ApiError('invalid username or password', 401)
+    except Exception as e:
+        raise ApiError(str(e), 500)
+
+    # Get email address from LDAP
+    if not email_verified:
+        try:
+            ldap_result = ldap_connection.search_s(userdn, ldap.SCOPE_SUBTREE, '(objectClass=*)', ['mail'])
+            email = ldap_result[0][1]['mail'][0].decode(sys.stdout.encoding)
```

Comment:
let's have a .lower() for the emails to avoid funny problems with capitalized emails in our LDAP

Reply:
modified